src/main/java/wooteco/subway/domain/Sections.java

```diff
@@ -0,0 +1,215 @@
+        }
+
+        return result;
+    }
+
+    public Section add(Section newSection) {
+        validateAddable(newSection);
+
+        if (getUpDestination().equals(newSection.getDownStation())) {
+            values.add(0, newSection);
+            return newSection;
+        }
+
+        if (getDownDestination().equals(newSection.getUpStation())) {
+            values.add(newSection);
+            return newSection;
+        }
+
+        for (Section section : values) {
+            if (section.hasSameUpStation(newSection)) {
+                return insertAfterUpStation(newSection, section);
+            }
+
+            if (section.hasSameDownStation(newSection)) {
+                return insertBeforeDownStation(newSection, section);
+            }
+        }
+        throw new IllegalStateException("구간을 추가할 수 있는 상태가 아닙니다.");
+    }
+
+    private Section insertAfterUpStation(Section newSection, Section section) {
+        int index = values.indexOf(section);
+        Section sectionToModify = new Section(section.getId(), newSection.getDownStation(), section.getDownStation(),
+            getNewDistance(section, newSection));
+        values.set(index, newSection);
+        values.add(index + 1, sectionToModify);
+        return sectionToModify;
+    }
+
+    private int getNewDistance(Section section, Section newSection) {
+        int newDistance = section.getDistance() - newSection.getDistance();
+        if (newDistance <= 0) {
+            throw new IllegalArgumentException("추가하려는 구간의 거리가 기존 구간의 거리보다 작아야합니다.");
+        }
+        return newDistance;
+    }
+
+    private Section insertBeforeDownStation(Section newSection, Section section) {
+        int index = values.indexOf(section);
+        Section sectionToModify = new Section(section.getId(), section.getUpStation(), newSection.getUpStation(),
+            getNewDistance(section, newSection));
+        values.set(index, sectionToModify);
+        values.add(index + 1, newSection);
+        return sectionToModify;
+    }
+
+    private void validateAddable(Section section) {
+        List<Section> foundSections = findSectionsOverlapped(section);
+
+        if (isInvalid(section, foundSections)) {
+            throw new IllegalArgumentException("이미 포함된 두 역을 가진 Section 을 추가할 수 없습니다.");
+        }
+    }
+
+    private List<Section> findSectionsOverlapped(Section newSection) {
+        return values.stream()
+            .filter(section ->
+                section.contains(newSection.getDownStation()) || section.contains(newSection.getUpStation()))
+            .collect(Collectors.toList());
+    }
+
+    private boolean isInvalid(Section section, List<Section> foundSections) {
+        return hasSameStations(section) ||
+            isInvalidCountInMiddle(foundSections) || isSameWithDestinations(foundSections);
+    }
+
+    private boolean hasSameStations(Section newSection) {
+        return values.stream().anyMatch(section -> section.hasSameStations(newSection));
+    }
```

Comment:
이 메서드가 지금은 새로운 구간을 추가하려 할 때 확인하는 용도로만 사용되고 있지만, 이 메서드 단독으로만 놓고 보면 파라미터 이름에 `new`가 들어갈 필요는 없어 보이는데 소주캉의 생각은 어떠신가요?
또한 구간"들"과 하나의 구간을 놓고 비교하는 것이니만큼 메서드 이름에는 `anyMatch`의 맥락을 드러낼 필요가 있어 보여요.

Reply:
그런데 다시 보니까, 모든 구간을 대상으로 `section.hasSameStations(newSection)` 확인을 할 필요가 있을까요? `validateAddable()`에서 `findSectionsOverlapped(section)`으로 찾은 구간들을 대상으로만 확인해도 되지 않을까요?

제가 잘못 이해한 부분이 있다면 코멘트 부탁드려요~

Reply:
그렇군요! 메서드 이름만 보면 현재 구간 목록에서 검증하는 것인지, 하나의 구간을 대상으로 검증하는 것인지 명확하지 않군요!!! 구간 목록에서 검증하는 것임을 밝히도록 메서드 명을 수정하고, `validateAddable()`에서 찾은 구간을 대상으로만 확인하도록 변경하였습니다. 개발을 진행할 때는 미처 체크하지 못했던 부분인데, 이렇게 꼼꼼하게 봐주셔서 정말 감사드립니다🙏🙏

```suggestion
   private boolean hasAnySectionWithSameStations(List<Section> foundSections,
        Section section) {
        return foundSections.stream()
            .anyMatch(it-> it.hasSameStations(section));
    }
```